src/dtls.c

```diff
@@ -290,60 +354,459 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+                ERROR_OUT(BUFFER_ERROR, dtls13_cleanup);
+            if ((len % 2) != 0)
+                ERROR_OUT(BUFFER_ERROR, dtls13_cleanup);
+            suites.hashSigAlgoSz = len;
+            XMEMCPY(suites.hashSigAlgo, tlsx.elements + OPAQUE16_LEN,
+                    len);
+        }
+
+        /* Supported groups */
+        ret = TlsxFindByType(&tlsx, TLSX_SUPPORTED_GROUPS,
+                             ch->extension);
+        if (ret != 0)
+            goto dtls13_cleanup;
+        if (tlsx.size != 0) {
+            ret = TLSX_SupportedCurve_Parse(ssl, tlsx.elements,
+                                         (word16)tlsx.size, 1, &parsedExts);
+            if (ret != 0)
+                goto dtls13_cleanup;
+        }
+
+        /* Key share */
+        ret = TlsxFindByType(&tlsx, TLSX_KEY_SHARE,
+                             ch->extension);
+        if (ret != 0)
+            goto dtls13_cleanup;
+        if (tlsx.size != 0) {
+            ret = TLSX_KeyShare_Parse_ClientHello(ssl, tlsx.elements,
+                                            (word16)tlsx.size, &parsedExts);
+            if (ret != 0)
+                goto dtls13_cleanup;
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+            haveKS = 1;
+#endif
+        }
+
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+        /* Pre-shared key */
+        ret = TlsxFindByType(&tlsx, TLSX_PRE_SHARED_KEY, ch->extension);
+        if (ret != 0)
+            goto dtls13_cleanup;
+        if (tlsx.size != 0) {
+            /* Let's just assume that the binders are correct here. We will
+             * actually verify this in the stateful part of the processing
+             * and if they don't match we will error out there anyway. */
+            byte modes;
+
+#ifndef NO_PSK
+            /* When we didn't find a valid ticket ask the user for the
+             * ciphersuite matching this identity */
+            if (!pskInfo->isValid) {
+                if (TLSX_PreSharedKey_Parse_ClientHello(&parsedExts,
+                        tlsx.elements, tlsx.size, ssl->heap) == 0)
+                    FindPskSuiteFromExt(ssl, parsedExts, pskInfo, &suites);
+                /* Revert to full handshake if PSK parsing failed */
+            }
+#endif
+
+            if (pskInfo->isValid) {
+                ret = TlsxFindByType(&tlsx, TLSX_PSK_KEY_EXCHANGE_MODES,
+                                     ch->extension);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+                if (tlsx.size == 0)
+                    ERROR_OUT(MISSING_HANDSHAKE_DATA, dtls13_cleanup);
+                ret = TLSX_PskKeyModes_Parse_Modes(tlsx.elements, tlsx.size,
+                                                  client_hello, &modes);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+                if ((modes & (1 << PSK_DHE_KE)) &&
+                        !ssl->options.noPskDheKe) {
+                    if (!haveKS)
```

Comment:
a zero sized keyshare extension shouldn't be fatal.

Reply:
Refactored to handle empty extensions but KeyShare should still have the two length bytes.